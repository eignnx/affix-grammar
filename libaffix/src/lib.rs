#[macro_use]
extern crate serde_derive;

pub mod fault;
pub mod gen;
pub mod literate;
pub mod parser;
pub mod semantics;

#[cfg(test)]
mod tests {
    use super::*;
    use std::convert::TryFrom;
    use std::fs;
    use std::path::PathBuf;

    static RNG_SEED: u64 = 123456;
    static MAX_VALUES: usize = 100;

    #[test]
    fn run_snapshots() {
        let inputs = PathBuf::from(std::env!("CARGO_MANIFEST_DIR"))
            .join("src")
            .join("snapshot_inputs");

        for dir_entry in fs::read_dir(inputs).unwrap() {
            let path = dir_entry.unwrap().path();
            let src_text = fs::read_to_string(&path).unwrap();
            let parse_res = parser::syntax::ParsedGrammar::try_from(&src_text[..]);

            let parse_snapshot_name = format!(
                "PARSE_SNAPSHOT__{}",
                path.file_stem().unwrap().to_string_lossy()
            );
            insta::assert_json_snapshot!(parse_snapshot_name, parse_res);

            if let Some(grammar) = parse_res.ok() {
                let generator = gen::Generator::new_seeded(grammar, RNG_SEED);
                let generated = generate_n(generator, MAX_VALUES);
                let gen_snapshot_name = format!(
                    "GEN_SNAPSHOT__{}",
                    path.file_stem().unwrap().to_string_lossy()
                );
                insta::assert_json_snapshot!(gen_snapshot_name, generated);
            }
        }
    }

    fn generate_n<R: rand::Rng>(
        mut gen: gen::Generator<R>,
        n: usize,
    ) -> fault::DynamicRes<Vec<String>> {
        let mut sentences = vec![];
        for _ in 0..n {
            match gen.generate() {
                Ok(sentence) => sentences.push(sentence),
                Err(fault::DynamicErr::MaxTrialsExceeded { .. }) => return Ok(sentences),
                Err(e) => return Err(e),
            }
        }
        Ok(sentences)
    }
}
