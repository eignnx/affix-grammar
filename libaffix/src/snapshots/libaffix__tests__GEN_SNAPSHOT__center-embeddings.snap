---
source: libaffix/src/lib.rs
expression: generated
---
{
  "Ok": [
    "an aardvark who an aardvark annoyed slept.",
    "a porcupine longed for the sea.",
    "the pterodactyl who the pterodactyl who a pterodactyl who a pterodactyl loved treated harshly in the workplace treated harshly in the workplace slept.",
    "an ostrich who an ostrich who the ostrich  the ostrich  the ostrich  the ostrich insulted complemented avoided plucked plucked killed the ostrich  an ostrich  an ostrich who an ostrich that the ostrich that an ostrich consumed intended to kill patronized avoided minimized.",
    "the dash hound longed for the sea.",
    "the porcupine  a porcupine was attracted to was murdered on the high seas.",
    "a porpoise visited a porpoise  the porpoise loved.",
    "a giraffe who the giraffe avoided longed for the sea.",
    "the dash hound visited a dash hound  the dash hound arrested.",
    "a cat killed the cat.",
    "a pterodactyl  the pterodactyl that the pterodactyl saw spat upon was murdered on the high seas.",
    "the pterodactyl that the pterodactyl  the pterodactyl  a pterodactyl was attracted to annoyed treated harshly in the workplace killed a pterodactyl.",
    "a sea turtle  a sea turtle plucked longed for the sea.",
    "the dog  the dog was attracted to slept.",
    "the giraffe who a giraffe that a giraffe  a giraffe that a giraffe  the giraffe that a giraffe consumed saw treated harshly in the workplace consumed had observed carefully patronized visited the giraffe.",
    "a pterodactyl visited the pterodactyl.",
    "a porpoise longed for the sea.",
    "the porcupine slept.",
    "a porcupine that the porcupine complemented was murdered on the high seas.",
    "a hamster who the hamster that a hamster  the hamster spat upon spat upon complemented slept.",
    "a porcupine that a porcupine that a porcupine arrested had observed carefully slept.",
    "the dog who a dog who a dog who the dog complemented arrested insulted killed a dog that the dog  the dog who a dog  a dog consumed minimized was attracted to saw.",
    "the dash hound that a dash hound  a dash hound that a dash hound had observed carefully plucked treated harshly in the workplace was murdered on the high seas.",
    "a pterodactyl that the pterodactyl that the pterodactyl who the pterodactyl who the pterodactyl avoided treated harshly in the workplace intended to kill arrested slept.",
    "an ostrich visited the ostrich  an ostrich who an ostrich insulted was attracted to.",
    "an ostrich visited an ostrich.",
    "an ostrich killed the ostrich  an ostrich who the ostrich  an ostrich who the ostrich who the ostrich who the ostrich avoided treated harshly in the workplace annoyed complemented spat upon loved.",
    "an ostrich  the ostrich annoyed visited an ostrich.",
    "an ostrich  the ostrich complemented was murdered on the high seas.",
    "the porpoise slept.",
    "a dash hound killed the dash hound.",
    "a giraffe  the giraffe that a giraffe that the giraffe saw was attracted to saw was murdered on the high seas.",
    "the aardvark visited an aardvark.",
    "an ostrich  an ostrich complemented longed for the sea.",
    "a dash hound killed a dash hound who the dash hound that the dash hound who the dash hound  a dash hound insulted intended to kill annoyed spat upon.",
    "the ostrich killed the ostrich.",
    "a hamster was murdered on the high seas.",
    "a cat longed for the sea.",
    "the cat visited a cat who a cat who a cat who the cat minimized saw arrested.",
    "a porpoise killed a porpoise.",
    "the aardvark longed for the sea.",
    "the porcupine  the porcupine  a porcupine consumed minimized killed the porcupine.",
    "the cat  a cat was attracted to killed the cat  a cat that a cat who a cat  the cat who the cat spat upon insulted minimized minimized annoyed.",
    "the hamster longed for the sea.",
    "an ostrich slept.",
    "a porpoise  the porpoise plucked slept.",
    "the sea turtle who the sea turtle  a sea turtle had observed carefully saw longed for the sea.",
    "a dash hound slept.",
    "a hamster  the hamster  a hamster  the hamster  a hamster minimized arrested minimized plucked slept.",
    "the dog killed the dog.",
    "an ostrich  an ostrich  the ostrich that the ostrich that the ostrich  an ostrich saw minimized patronized arrested patronized visited an ostrich that an ostrich that the ostrich that the ostrich that an ostrich plucked saw avoided spat upon.",
    "the dog who a dog that the dog that the dog plucked consumed patronized visited the dog  a dog plucked.",
    "the dash hound slept.",
    "a dog longed for the sea.",
    "the sea turtle that a sea turtle patronized visited the sea turtle that the sea turtle intended to kill.",
    "a dog visited the dog who a dog  a dog saw was attracted to.",
    "the porpoise  a porpoise who the porpoise intended to kill saw slept.",
    "the pterodactyl visited a pterodactyl  a pterodactyl intended to kill.",
    "the dog who the dog saw visited the dog.",
    "the porpoise  a porpoise had observed carefully killed the porpoise who a porpoise  a porpoise had observed carefully consumed.",
    "the pterodactyl was murdered on the high seas.",
    "the sea turtle visited the sea turtle.",
    "a porcupine who the porcupine who the porcupine  the porcupine  the porcupine consumed was attracted to intended to kill avoided longed for the sea.",
    "the dog that the dog that a dog  a dog that a dog  the dog that a dog who the dog annoyed insulted annoyed intended to kill plucked saw intended to kill killed a dog  a dog who the dog  a dog complemented was attracted to treated harshly in the workplace.",
    "a cat who a cat had observed carefully longed for the sea.",
    "the cat who a cat intended to kill slept.",
    "the porpoise killed a porpoise  the porpoise  the porpoise insulted saw.",
    "the porcupine that the porcupine who the porcupine annoyed plucked slept.",
    "the dog longed for the sea.",
    "a dash hound was murdered on the high seas.",
    "the sea turtle that the sea turtle annoyed visited a sea turtle.",
    "a cat who the cat consumed was murdered on the high seas.",
    "the sea turtle slept.",
    "the porcupine that a porcupine avoided killed a porcupine  a porcupine was attracted to.",
    "a porpoise  a porpoise who the porpoise  the porpoise had observed carefully arrested arrested killed the porpoise that a porpoise loved.",
    "a dash hound that the dash hound insulted slept.",
    "a hamster  a hamster spat upon slept.",
    "the hamster who a hamster who a hamster  a hamster  a hamster that a hamster that the hamster loved plucked avoided loved patronized was attracted to longed for the sea.",
    "a cat was murdered on the high seas.",
    "the aardvark visited the aardvark that an aardvark spat upon.",
    "the pterodactyl that a pterodactyl loved slept.",
    "the porcupine longed for the sea.",
    "the aardvark slept.",
    "a pterodactyl slept.",
    "the dash hound that a dash hound complemented longed for the sea.",
    "the giraffe was murdered on the high seas.",
    "the sea turtle longed for the sea.",
    "a sea turtle that a sea turtle who the sea turtle that a sea turtle who the sea turtle that the sea turtle  a sea turtle who a sea turtle spat upon consumed complemented saw had observed carefully was attracted to saw slept.",
    "the dog visited a dog.",
    "the pterodactyl slept.",
    "a porpoise who a porpoise  a porpoise consumed arrested slept.",
    "the cat who the cat consumed slept.",
    "a porcupine who the porcupine arrested was murdered on the high seas.",
    "an ostrich killed the ostrich.",
    "a dog  the dog loved was murdered on the high seas.",
    "the dog that the dog arrested visited the dog that a dog that a dog who a dog insulted consumed treated harshly in the workplace.",
    "an ostrich who the ostrich plucked visited an ostrich  the ostrich that the ostrich minimized plucked.",
    "the giraffe slept.",
    "a pterodactyl was murdered on the high seas.",
    "the dog that a dog annoyed longed for the sea."
  ]
}
